### PR TITLE
UITreeView (& test method conventions)

### PR DESCRIPTION
**Pull request by rzats**
### Contains

- A core UITreeView widget as described in [rzats/gsoc-nui-editor#1](https://github.com/rzats/gsoc-nui-editor/issues/1)!

image

- A `Tree` data structure that contains the widget's elements and a `TreeModel` helper class.

- `TreeViewTestScreen` - fairly self-explanatory :stuck_out_tongue_winking_eye:

- A codestyle fix renaming all the test methods to `testMethodName` as opposed to `methodName` - this is a mostly unrelated change as suggested in a Slack discussion.

_This PR was previously located in my Terasology fork. In retrospect, this wasn't a great idea: local PRs have no @GooeyHub testing, no mergeability checks with the upstream repository and little to none community exposure: IMO, going for a base repository fork is a better idea for subsequent deliverables._

### How to test

- Launch the game and type `showScreen treeViewTestScreen` in the in-game console. Make sure you do so manually without using autocompletion - otherwise #2329 triggers, causing initialization issues.
- Play around with the trees!

### Controls

- Mouse wheel: scroll up/down.
- LMB: select/deselect a node.
- RMB: expand/collapse a node.
- Delete: remove a node (and all its' children).
- Up/Down: change a node's position within the parent node.
- Ctrl+C: copy a selected node.
- Ctrl+V: paste the copied node as a child of the currently selected node.
- Ctrl+A / Insert: add a new child with a placeholder value to the currently selected node.

### CODE REVIEW

**Comment by GooeyHub** [comment]
Hooray Jenkins reported success with all tests good!

**Comment by GooeyHub** [comment]
Hooray Jenkins reported success with all tests good!

**Comment by GooeyHub** [comment]
Hooray Jenkins reported success with all tests good!

**Comment by GooeyHub** [comment]
Hooray Jenkins reported success with all tests good!

**Comment by GooeyHub** [comment]
Hooray Jenkins reported success with all tests good!

File: engine/src/main/java/org/terasology/rendering/nui/widgets/models/Tree.java [hunk 1]
```diff
@@ -0,0 +1,315 @@
...
+        return this.children;
+    }
+
+
+    /**
+     * @param node The {@code Tree}the index of which is to be returned.
+     * @return The index of the specified {@code Tree}.
+     */
+    public int getIndex(Tree<T> node) {
+        if (node == null) {
```
Comment (msteiger) [inline engine/src/main/java/org/terasology/rendering/nui/widgets/models/Tree.java]: You can use `Preconditions.checkArgument()` here, if you like.

**Comment by msteiger** [comment]
I looked over it briefly and didn't test all corner cases, but afaict, it looks good.

Slightly offtopic: The behavior node editor currently uses `org.abego.treelayout.TreeLayout` to organize tree structured UI elements on the screen. Does it make sense to connect it with the code in this PR or can we consider those two as completely separate things?

**Comment by rzats** [comment]
@msteiger: thanks for the test!

I've actually checked the BTE package before starting work on this PR - it's specifically meant for [tree visualizations](https://camo.githubusercontent.com/bba1a2474558e8ab3715317163ca87163129fa36/687474703a2f2f747265656c61796f75742e736f75726365666f7267652e6e65742f696d6167652f547265654772617068566965772d466967757265372e706e67) as opposed to [tree views](https://en.wikipedia.org/wiki/Tree_view), so not a great fit.
`Tree` could theoretically be modified to implement [`org.abego.treelayout.TreeForTreeLayout`](https://github.com/abego/treelayout/blob/master/org.abego.treelayout/src/main/java/org/abego/treelayout/TreeForTreeLayout.java), although I don't see any benefits from doing so.

**Comment by msteiger** [comment]
That's what I expected. Thanks for the clarification.

File: engine/src/main/java/org/terasology/rendering/nui/widgets/models/Tree.java (hunk 1 above)
Comment (rzats) [inline engine/src/main/java/org/terasology/rendering/nui/widgets/models/Tree.java]: Just noticed the notification for this comment. Fixed all the checks accordingly.

**Comment by GooeyHub** [comment]
Hooray Jenkins reported success with all tests good!

**Comment by flo** [comment]
@rzats got a IndexOutOfBoundException while testing it:
http://pastebin.com/hGHbfPvH

**Comment by flo** [comment]
The crash happens when you minimize a node near the end of the list while a child node is still selected. When it isn't near the end of the list, another item gets selected (which is also strange).

**Comment by flo** [comment]
Great work otherwise, the PR can be merged once the bug is fixed.

**Comment by flo** [comment]
Merged, thanks for the PR

**Comment by GooeyHub** [comment]
Hooray Jenkins reported success with all tests good!
